File: behavioral/typescript/command_structured/concrete_commands.ts
import { Command } from './interfaces';
import { Light, Stereo, BankAccount } from './receivers';

export class LightOnCommand implements Command {
  private light: Light;
  
  constructor(light: Light) {
    this.light = light;
  }
  
  execute(): void {
    this.light.on();
  }
  
  undo(): void {
    this.light.off();
  }
}

export class LightOffCommand implements Command {
  private light: Light;
  
  constructor(light: Light) {
    this.light = light;
  }
  
  execute(): void {
    this.light.off();
  }
  
  undo(): void {
    this.light.on();
  }
}

export class StereoOnCommand implements Command {
  private stereo: Stereo;
  private previousVolume: number = 0;
  
  constructor(stereo: Stereo) {
    this.stereo = stereo;
  }
  
  execute(): void {
    this.stereo.on();
    this.stereo.setVolume(10);
    this.previousVolume = 10;
  }
  
  undo(): void {
    this.stereo.setVolume(this.previousVolume);
    this.stereo.off();
  }
}

export class StereoOffCommand implements Command {
  private stereo: Stereo;
  
  constructor(stereo: Stereo) {
    this.stereo = stereo;
  }
  
  execute(): void {
    this.stereo.off();
  }
  
  undo(): void {
    this.stereo.on();
  }
}

export class DepositCommand implements Command {
  private account: BankAccount;
  private amount: number;
  
  constructor(account: BankAccount, amount: number) {
    this.account = account;
    this.amount = amount;
  }
  
  execute(): void {
    this.account.deposit(this.amount);
  }
  
  undo(): void {
    this.account.withdraw(this.amount);
  }
}

export class WithdrawCommand implements Command {
  private account: BankAccount;
  private amount: number;
  private success: boolean = false;
  
  constructor(account: BankAccount, amount: number) {
    this.account = account;
    this.amount = amount;
  }
  
  execute(): void {
    const oldBalance = this.account.getBalance();
    this.account.withdraw(this.amount);
    this.success = this.account.getBalance() < oldBalance;
  }
  
  undo(): void {
    if (this.success) {
      this.account.deposit(this.amount);
    }
  }
}

File: behavioral/dart/command_structured/concrete_commands.dart
import 'interfaces.dart';
import 'receivers.dart';

class LightOnCommand extends Command {
  final Light _light;
  
  LightOnCommand(this._light);
  
  @override
  void execute() {
    _light.on();
  }
  
  @override
  void undo() {
    _light.off();
  }
}

class LightOffCommand extends Command {
  final Light _light;
  
  LightOffCommand(this._light);
  
  @override
  void execute() {
    _light.off();
  }
  
  @override
  void undo() {
    _light.on();
  }
}

class StereoOnCommand extends Command {
  final Stereo _stereo;
  int _previousVolume = 0;
  
  StereoOnCommand(this._stereo);
  
  @override
  void execute() {
    _stereo.on();
    _stereo.setVolume(10);
    _previousVolume = 10;
  }
  
  @override
  void undo() {
    _stereo.setVolume(_previousVolume);
    _stereo.off();
  }
}

class StereoOffCommand extends Command {
  final Stereo _stereo;
  
  StereoOffCommand(this._stereo);
  
  @override
  void execute() {
    _stereo.off();
  }
  
  @override
  void undo() {
    _stereo.on();
  }
}

class DepositCommand extends Command {
  final BankAccount _account;
  final double _amount;
  
  DepositCommand(this._account, this._amount);
  
  @override
  void execute() {
    _account.deposit(_amount);
  }
  
  @override
  void undo() {
    _account.withdraw(_amount);
  }
}

class WithdrawCommand extends Command {
  final BankAccount _account;
  final double _amount;
  bool _success = false;
  
  WithdrawCommand(this._account, this._amount);
  
  @override
  void execute() {
    final oldBalance = _account.getBalance();
    _account.withdraw(_amount);
    _success = _account.getBalance() < oldBalance;
  }
  
  @override
  void undo() {
    if (_success) {
      _account.deposit(_amount);
    }
  }
}
